_studio/shared/umc/codec/jpeg_enc/src/jpegenc.cpp

```diff
@@ -0,0 +1,5025 @@
+
+  TRC1("  emit marker ",JM_APP0);
+  TRC1("    length ",len);
+
+  jerr = m_BitStreamOut.WriteWord(0xff00 | JM_APP0);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteWord(len);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // identificator JFIF
+  jerr = m_BitStreamOut.WriteByte('J');
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteByte('F');
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteByte('I');
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteByte('F');
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteByte(0);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // version
+  jerr = m_BitStreamOut.WriteWord(0x0102);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // units: 0 - none, 1 - dot per inch, 2 - dot per cm
+  jerr = m_BitStreamOut.WriteByte(m_jfif_app0_units);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // xDensity
+  jerr = m_BitStreamOut.WriteWord(m_jfif_app0_xDensity);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // yDensity
+  jerr = m_BitStreamOut.WriteWord(m_jfif_app0_yDensity);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // xThumbnails, yThumbnails
+  jerr = m_BitStreamOut.WriteByte(0);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteByte(0);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  return JPEG_OK;
+} // CJPEGEncoder::WriteAPP0()
+
+
+JERRCODE CJPEGEncoder::WriteAPP14(void)
+{
+  int len;
+  JERRCODE jerr;
+
+  TRC0("-> WriteAPP14");
+
+  len = 2 + 5 + 2 + 2 + 2 + 1;
+
+  TRC1("  emit marker ",JM_APP14);
+  TRC1("    length ",len);
+
+  jerr = m_BitStreamOut.WriteWord(0xff00 | JM_APP14);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteWord(len);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // identificator Adobe
+  jerr = m_BitStreamOut.WriteByte('A');
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteByte('d');
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteByte('o');
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteByte('b');
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  jerr = m_BitStreamOut.WriteByte('e');
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // version
+  jerr = m_BitStreamOut.WriteWord(100);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // Flags 0
+  jerr = m_BitStreamOut.WriteWord(0);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  // Flags 1
+  jerr = m_BitStreamOut.WriteWord(0);
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  switch(m_jpeg_color)
+  {
+  case JC_YCBCR:
+    jerr = m_BitStreamOut.WriteByte(1);
+    break;
+
+  case JC_YCCK:
+    jerr = m_BitStreamOut.WriteByte(2);
+    break;
+
+  default:
+    jerr = m_BitStreamOut.WriteByte(0);
+    break;
+  }
+
+  if(JPEG_OK != jerr)
+    return jerr;
+
+  return JPEG_OK;
+} // CJPEGEncoder::WriteAPP14()
+
+
+JERRCODE CJPEGEncoder::WriteCOM(
+  char*  comment)
+{
+  std::string buf;
```

Comment:
I switched to std::string here, hope this is OK. Code `sz = (int)MFX_MIN(strnlen_s(comment, 127),127);` from previous version looks suspicious.

Reply:
much better, thank you. Don't forget to move this patch to internal repo.